### PR TITLE
Feature/encounters

### PR DESCRIPTION
**Pull request by edcohen08**
Base functionality for CRUD on encounters via either a patient or a user

### CODE REVIEW

File: welkin/models/encounter.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            patient_id = self.patientId
+        return self._client.Disposition(encounter_id=self.id, patient_id=patient_id)
+
+
+class Encounters(Collection):
+    resource = Encounter
+    patient_id: str = None
+    user_id: str = None
+
+    def __init__(self, patient_id=None, user_id=None):
```
Comment (samamorgan) [inline welkin/models/encounter.py:161]: I don't think we can add a constructor. If we do, objects can't be serialized to this class from JSON responses.

File: welkin/models/encounter.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    @property
+    def Disposition(self):
+        if getattr(self, "patient_id"):
+            patient_id = self.patient_id
+        else:
+            patient_id = self.patientId
+        return self._client.Disposition(encounter_id=self.id, patient_id=patient_id)
+
+
+class Encounters(Collection):
```
Comment (samamorgan) [inline welkin/models/encounter.py:156]: The changes to this class change the overall behavior of this collection vs others. I understand why you went this way (so you can use class properties of parent resources), but I think it's likely best we step back and just use a getter method for subresources so we can keep behavior consistent between APIs.

File: test/test_encounters.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    end = start + timedelta(hours=1)
+
+    data = {
+        "title": "new enc",
+        "description": "new encounter",
+        "templateName": "etmp-encounter1",
+        "currentScheduledAppointment": {
+            "eventType": "ENCOUNTER",
+            "eventTitle": "NEW ENC",
+            "startDateTime": clean_datetime(start),
```
Comment (samamorgan) [inline test/test_encounters.py:26]: Don't use the `clean_datetime` directly, it's called as part of `Client.prepare_request`

File: test/test_encounters.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                {
+                    "participantId": "a9618392-799d-4664-a896-5f1756f8d336",
+                    "participantRole": "psm",
+                },
+                {"participantId": patient.id, "participantRole": "patient"},
+            ],
+        },
+    }
+
+    encounter = patient.Encounter
```
Comment (samamorgan) [inline test/test_encounters.py:39]: I'd do `encounter = patient.Encounter(**data)`. Resource subclasses dict, which takes key/value pairs as initialization arguments.

File: test/test_encounters.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    assert isinstance(encounters, Encounters)
+    assert len(encounters) == 8
+    assert isinstance(encounters[0], Encounter)
+    assert len(vcr_cassette) == 1
+
+
+@pytest.mark.vcr()
+def test_all_encounters_user_read(client, vcr_cassette):
+    user = client.User(id="a9618392-799d-4664-a896-5f1756f8d336")
+    encounters = user.Encounters.get(params={"withCareTeam": "false"})
```
Comment (samamorgan) [inline test/test_encounters.py:64]: I'd recommend adding parameters as method arguments: `user.Encounters().get(with_care_team=False)`

**Review by samamorgan — CHANGES_REQUESTED** [review]
Lots of good work here! Added quite a few comments, mostly so we can stay consistent across APIs. Also, before approval, we need to ensure code coverage is as good or better than the last high water mark.

File: welkin/models/patient.py [hunk 6]
```diff
@@ -17,6 +20,14 @@ def delete(self):
     def __str__(self):
         return f"{self.firstName} {self.lastName}"
 
+    @property
+    def Encounter(self):
+        return self._client.Encounter(patient_id=self.id)
```
Comment (samamorgan) [inline welkin/models/patient.py:25]: If we continue to use the property decorator here, this should return an uninitialized class.

```python
@property
def Encounter(self):
    encounter = self._client.Encounter
    encounter.patient_id = self.id

    return encounter
```

File: test/test_encounters.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                    "participantRole": "psm",
+                },
+                {"participantId": patient.id, "participantRole": "patient"},
+            ],
+        },
+    }
+
+    encounter = patient.Encounter(**data)
+
+    created_enc = encounter.create()
```
Comment (samamorgan) [inline test/test_encounters.py:39]: Just an FYI: You don't need to capture the output of `encounter.create()`. Calling `Resource.create()` makes the api call, updates `self`, then returns `self`. This is so you can use the method on an existing instance or create an instance and make an API call in one line:

```python
resource = welkin.Resource(id="foo")
resource.create()

# OR

resource = welkin.Resource(id="foo").create()
```

**Review by samamorgan — APPROVED** [review]
LGTM!

File: test/test_encounters.py (hunk 7 above)
Comment (edcohen08) [reply test/test_encounters.py:39]: ah good to know, thank you

**Pull request merged**